app/main.py

```diff
@@ -84,131 +1,129 @@
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
-
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
-
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
-
-    # arthur
-    arthur = knightsConfig["arthur"]
-
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
-
-    # apply weapon
-    arthur["power"] += arthur["weapon"]["power"]
 
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
+    def attack(self, other):
+        damage = max(0, self.final_stats["power"] - other.final_stats["protection"])
+        other.final_stats["hp"] -= damage
+        return damage
 
-        if "protection" in arthur["potion"]["effect"]:
-            arthur["protection"] += arthur["potion"]["effect"]["protection"]
 
-        if "hp" in arthur["potion"]["effect"]:
-            arthur["hp"] += arthur["potion"]["effect"]["hp"]
+def simulate_battle(knight1, knight2):
 
-    # mordred
-    mordred = knightsConfig["mordred"]
+    round_counter = 1
+    while knight1.final_stats["hp"] > 0 and knight2.final_stats["hp"] > 0:
+        print(f"Round {round_counter}: {knight1.name} attacks {knight2.name}!")
+        damage = knight1.attack(knight2)
+        print(f"{knight1.name} deals {damage} damage. {knight2.name} has {knight2.final_stats['hp']} HP left.")
 
-    # apply armour
-    mordred["protection"] = 0
-    for a in mordred["armour"]:
-        mordred["protection"] += a["protection"]
+        if knight2.final_stats["hp"] <= 0:
+            print(f"{knight1.name} wins!")
+            return {"winner": knight1.name, "loser": knight2.name}
 
-    # apply weapon
-    mordred["power"] += mordred["weapon"]["power"]
+        print(f"Round {round_counter}: {knight2.name} attacks {knight1.name}!")
+        damage = knight2.attack(knight1)
+        print(f"{knight2.name} deals {damage} damage. {knight1.name} has {knight1.final_stats['hp']} HP left.")
 
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
+        if knight1.final_stats["hp"] <= 0:
+            print(f"{knight2.name} wins!")
+            return {"winner": knight2.name, "loser": knight1.name}
 
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
+        round_counter += 1
 
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
+    return {"winner": None, "loser": None}  # У разі нічиєї
```

Comment:
The battle simulation logic is repetitive. Consider refactoring this section to reduce redundancy, possibly by creating a helper function to handle the attack sequence for each knight.